### PR TITLE
BUG : upgrade clone_document_root

### LINKED ISSUE #1518: Can't add outline
Outlines can no longer be added in version 3 or higher. The generated file has no bookmarks.

## Environment

Windows 11

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfReader, PdfWriter

writer = PdfWriter()
reader = PdfReader(open("test.pdf", "rb"))

# copy original pdf
writer.clone_document_from_reader(reader)

# add bookmark
writer.add_outline_item("Page 1", 0)
writer.add_outline_item("Page 2", 1)

# generate pdf
outputFileName = "test_bookmarked.pdf"
with open(outputFileName, "wb") as o:
    writer.write(o)

```

This results in the following PDF:

[test_bookmarked.pdf](https://github.com/py-pdf/pypdf/files/10303954/test_bookmarked.pdf)

As you can see, the PDF does not have bookmarks. It should have bookmarks.

### PR DESCRIPTION
**Pull request by pubpub-zz**
fixes #1518

### CODE REVIEW

File: pypdf/_writer.py [hunk 1]
```diff
@@ -800,7 +800,14 @@ def clone_reader_document_root(self, reader: PdfReader) -> None:
 
         :param reader:  PdfReader from the document root should be copied.
         """
-        self._root_object = cast(DictionaryObject, reader.trailer[TK.ROOT])
+        self._root_object = cast(DictionaryObject, reader.trailer[TK.ROOT].clone(self))
+        self._root = self._root_object.indirect_reference
```
Comment (MartinThoma) [inline pypdf/_writer.py:804]:
```suggestion
        self._root = self._root_object.indirect_reference  # type: ignore[assignment]
```

File: pypdf/_writer.py [hunk 2]
```diff
@@ -800,7 +800,14 @@ def clone_reader_document_root(self, reader: PdfReader) -> None:
...
+        self._pages = self._root_object.raw_get("/Pages")
+        self._flatten()
+        self._root_object[NameObject("/Pages")][
```
Comment (MartinThoma) [inline pypdf/_writer.py:807]:
```suggestion
        self._root_object[NameObject("/Pages")][  # type: ignore[index]
```

File: pypdf/_writer.py [hunk 3]
```diff
@@ -813,6 +820,51 @@ def cloneReaderDocumentRoot(self, reader: PdfReader) -> None:  # pragma: no cove
...
+            NameObject(PG.ROTATE),
+        )
+        if inherit is None:
+            inherit = {}
+        if pages is None:
+            # Fix issue 327: set flattened_pages attribute only for
+            # decrypted file
+            pages = self._root_object["/Pages"].get_object()  # type: ignore
+            self.flattened_pages = ArrayObject()
+
```
Comment (MartinThoma) [inline pypdf/_writer.py:842]:
```suggestion
assert pages is not None  # hint for mypy
```

File: pypdf/_writer.py [hunk 4]
```diff
@@ -813,6 +820,51 @@ def cloneReaderDocumentRoot(self, reader: PdfReader) -> None:  # pragma: no cove
...
+                if isinstance(page, IndirectObject):
+                    addt["indirect_reference"] = page
+                self._flatten(page.get_object(), inherit, **addt)
+        elif t == "/Page":
+            for attr_in, value in list(inherit.items()):
+                # if the page has it's own value, it does not inherit the
+                # parent's value:
+                if attr_in not in pages:
+                    pages[attr_in] = value
+            page_obj = PageObject(self, indirect_reference)
```
Comment (MartinThoma) [inline pypdf/_writer.py:862]:
```suggestion
            page_obj = PageObject(self, indirect_reference)  # type: ignore
```

File: pypdf/_writer.py [hunk 5]
```diff
@@ -813,6 +820,51 @@ def cloneReaderDocumentRoot(self, reader: PdfReader) -> None:  # pragma: no cove
...
+            page_obj.update(pages)
```
Comment (MartinThoma) [inline pypdf/_writer.py:863]:
```suggestion
            page_obj.update(pages)  # type: ignore
```

File: pypdf/_writer.py [hunk 6]
```diff
@@ -829,9 +881,8 @@ def clone_document_from_reader(
             (delegates to append_pages_from_reader). The single parameter of the
             callback is a reference to the page just appended to the document.
         """
-        # TODO : ppZZ may be limited because we do not copy all info...
         self.clone_reader_document_root(reader)
-        self.append_pages_from_reader(reader, after_page_append)
+        self._info = reader.trailer[TK.INFO].clone(self).indirect_reference
```
Comment (MartinThoma) [inline pypdf/_writer.py:885]:
```suggestion
        self._info = reader.trailer[TK.INFO].clone(self).indirect_reference  # type: ignore
```

File: pypdf/_writer.py [hunk 7]
```diff
@@ -813,6 +820,51 @@ def cloneReaderDocumentRoot(self, reader: PdfReader) -> None:  # pragma: no cove
...
+        )
+        if inherit is None:
+            inherit = {}
+        if pages is None:
+            # Fix issue 327: set flattened_pages attribute only for
+            # decrypted file
+            pages = self._root_object["/Pages"].get_object()  # type: ignore
+            self.flattened_pages = ArrayObject()
+assert pages is not None  # hint for mypy
+        t = "/Pages"
```
Comment (MartinThoma) [inline pypdf/_writer.py:843]:
```suggestion
        assert pages is not None  # hint for mypy
        t = "/Pages"
```

**Comment by pubpub-zz** [comment]
You've been too fast to apply the fixes.😊
I've been a little confused during fixing: `clone_reader_document_root` not only clones the root but all the info below, including pages : the `after_page_append` param in  `clone_document_from_reader` is now meaning less : Should we add some code to declare it obsolete or just apply a for loop on all pages ?
Also the only difference difference is now the copy of the info structure that is copied...

**Comment by MartinThoma** [comment]
> You've been too fast to apply the fixes.blush

Thank you for being so quick to create the PR :hugs: I was hoping I can support a bit with the tedious mypy issues :-)

**Comment by MartinThoma** [comment]
> Should we add some code to declare it obsolete

What do you want to declare obsolete?

I want to keep things a bit more stable in 2023; we had a lot of changes in 2022 and we need to be careful not to do too much. I don't want to loose trust of users.

**Comment by MartinThoma** [comment]
> Also the only difference difference is now the copy of the info structure that is copied...

Could you adjust one / both of the docstrings to mention this?

**Comment by MartinThoma** [comment]
I'll make a release with the fix today :-) (potentially also a "backport" to PyPDF2)

**Comment by pubpub-zz** [comment]
> > Should we add some code to declare it obsolete
> 
> What do you want to declare obsolete?
> 
> I want to keep things a bit more stable in 2023; we had a lot of changes in 2022 and we need to be careful not to do too much. I don't want to loose trust of users.

The parameter `clone_document_from_reader` which is currently unused

**Comment by MartinThoma** [comment]
The method `cloneDocumentFromReader` exists for a long time (before https://github.com/py-pdf/pypdf/blame/e48bc6d156cca69c8e7f411ec0de7967bcf63eed/PyPDF2/_writer.py). We can add a hint to point people to the new method, but I would not remove it for the moment. Simply for the reason to keep the interface stable for a while

**Comment by pubpub-zz** [comment]
> The method `cloneDocumentFromReader` exists for a long time (before https://github.com/py-pdf/pypdf/blame/e48bc6d156cca69c8e7f411ec0de7967bcf63eed/PyPDF2/_writer.py). We can add a hint to point people to the new method, but I would not remove it for the moment. Simply for the reason to keep the interface stable for a while

I dislike the idea of  removing this function too. Will add comments immediately

**Comment by pubpub-zz** [comment]
@MartinThoma 
stdby I may have detected a late bug

**Comment by pubpub-zz** [comment]
@MartinThoma 
stdby I may have detected a late bug

**Comment by pubpub-zz** [comment]
issue is now fixed. Quite tricky : files were raising issues with Acrobat Reader but not on Sumatra nor pdf.js.
issue located with https://www.pdf-online.com/osa/repair.aspx

**Comment by pubpub-zz** [comment]
all yours

**Comment by MartinThoma** [comment]
Thank you :pray: 

I'll make a release this year (at least in my timezone xD); but I need some sleep before I mess something up :-)

**Pull request merged**